File: examples/wasm_plugin/plugin/fib.go
// examples/wasm_plugin/plugin/fib.go
// A WebAssembly plugin for Fibonacci calculation.
//
// IMPORTANT: This requires TinyGo, not standard Go!
// Standard Go's wasip1 target does NOT support plugin-style WASM modules.
//
// Build with TinyGo:
//   tinygo build -o fib.wasm -target=wasi fib.go
//
// Why TinyGo is required:
// - Standard Go runs _start (main) and exits, closing the module
// - TinyGo's architecture allows exported functions to remain callable
package main

import (
	"unsafe"
)

// Memory allocation for WASM
var memory []byte

//export alloc
func alloc(size uint32) uint32 {
	offset := uint32(len(memory))
	// Ensure 8-byte alignment for int64
	if offset%8 != 0 {
		offset += 8 - offset%8
	}
	newLen := offset + size
	if uint32(cap(memory)) < newLen {
		newMem := make([]byte, newLen*2)
		copy(newMem, memory)
		memory = newMem[:newLen]
	} else {
		memory = memory[:newLen]
	}
	return offset
}

//export plugin_name
func pluginName() (ptr uint32, size uint32) {
	name := "fib"
	ptr = alloc(uint32(len(name)))
	copy(memory[ptr:], name)
	return ptr, uint32(len(name))
}

//export plugin_version
func pluginVersion() (ptr uint32, size uint32) {
	version := "1.0.0-wasm"
	ptr = alloc(uint32(len(version)))
	copy(memory[ptr:], version)
	return ptr, uint32(len(version))
}

//export call_fast
func fibFast(n int64) int64 {
	if n <= 1 {
		return n
	}
	a, b := int64(0), int64(1)
	for i := int64(2); i <= n; i++ {
		a, b = b, a+b
	}
	return b
}

//export call_matrix
func fibMatrix(n int64) int64 {
	if n <= 1 {
		return n
	}

	// Matrix multiplication
	mul := func(a, b [2][2]int64) [2][2]int64 {
		return [2][2]int64{
			{a[0][0]*b[0][0] + a[0][1]*b[1][0], a[0][0]*b[0][1] + a[0][1]*b[1][1]},
			{a[1][0]*b[0][0] + a[1][1]*b[1][0], a[1][0]*b[0][1] + a[1][1]*b[1][1]},
		}
	}

	// Fast power
	result := [2][2]int64{{1, 0}, {0, 1}}
	base := [2][2]int64{{1, 1}, {1, 0}}

	for n > 0 {
		if n&1 == 1 {
			result = mul(result, base)
		}
		base = mul(base, base)
		n >>= 1
	}

	return result[0][1]
}

//export call_isFib
func isFib(n int64) int32 {
	if n < 0 {
		return 0
	}

	// Check if 5*n^2+4 or 5*n^2-4 is a perfect square
	check := func(x int64) bool {
		sqrt := int64(0)
		for sqrt*sqrt < x {
			sqrt++
		}
		return sqrt*sqrt == x
	}

	n2 := n * n
	if check(5*n2+4) || check(5*n2-4) {
		return 1
	}
	return 0
}

//export call_range_
func fibRange(n int64) (ptr uint32, count uint32) {
	if n < 0 {
		return 0, 0
	}

	count = uint32(n + 1)
	size := count * 8
	ptr = alloc(size)

	// Fill array with Fibonacci numbers
	a, b := int64(0), int64(1)
	for i := int64(0); i <= n; i++ {
		offset := ptr + uint32(i*8)
		*(*int64)(unsafe.Pointer(uintptr(offset))) = a
		a, b = b, a+b
	}

	return ptr, count
}

func main() {
	// main is required but not used for TinyGo WASM plugins
}

File: examples/wasm_plugin/plugin/fib.ts
// fib.ts - Fibonacci WASM plugin in AssemblyScript
// Build: asc fib.ts -o fib.wasm --optimize --runtime stub --initialMemory 2

// Heap starts at 64KB (plenty of room within 2 pages = 128KB)
var heapPtr: usize = 65536;

// Memory allocator
export function alloc(size: u32): u32 {
    if (heapPtr % 8 != 0) {
        heapPtr += 8 - (heapPtr % 8);
    }
    const result = heapPtr as u32;
    heapPtr += size as usize;
    return result;
}

// Plugin name
export function plugin_name(resultPtr: u32): void {
    const name = "fib";
    const len = 3;
    const ptr = alloc(len);

    for (let i = 0; i < len; i++) {
        store<u8>(ptr + i, name.charCodeAt(i));
    }

    store<u32>(resultPtr, ptr);
    store<u32>(resultPtr + 4, len);
}

// Plugin version
export function plugin_version(resultPtr: u32): void {
    const version = "1.0.0-as";
    const len = 8;
    const ptr = alloc(len);

    for (let i = 0; i < len; i++) {
        store<u8>(ptr + i, version.charCodeAt(i));
    }

    store<u32>(resultPtr, ptr);
    store<u32>(resultPtr + 4, len);
}

// Fibonacci - O(n)
export function call_fast(n: i64): i64 {
    if (n <= 1) return n;
    let a: i64 = 0;
    let b: i64 = 1;
    for (let i: i64 = 2; i <= n; i++) {
        const tmp = a + b;
        a = b;
        b = tmp;
    }
    return b;
}

// Fibonacci - O(log n)
export function call_matrix(n: i64): i64 {
    if (n <= 1) return n;

    var r00: i64 = 1, r01: i64 = 0, r10: i64 = 0, r11: i64 = 1;
    var b00: i64 = 1, b01: i64 = 1, b10: i64 = 1, b11: i64 = 0;
    var m = n;

    while (m > 0) {
        if (m & 1) {
            const t00 = r00 * b00 + r01 * b10;
            const t01 = r00 * b01 + r01 * b11;
            const t10 = r10 * b00 + r11 * b10;
            const t11 = r10 * b01 + r11 * b11;
            r00 = t00; r01 = t01; r10 = t10; r11 = t11;
        }
        const t00 = b00 * b00 + b01 * b10;
        const t01 = b00 * b01 + b01 * b11;
        const t10 = b10 * b00 + b11 * b10;
        const t11 = b10 * b01 + b11 * b11;
        b00 = t00; b01 = t01; b10 = t10; b11 = t11;
        m >>= 1;
    }
    return r01;
}

// Check if Fibonacci
export function call_isFib(n: i64): i32 {
    if (n < 0) return 0;
    const n2 = n * n;
    const c1 = 5 * n2 + 4;
    const c2 = 5 * n2 - 4;

    var s1: i64 = 0;
    while (s1 * s1 < c1) s1++;
    if (s1 * s1 == c1) return 1;

    var s2: i64 = 0;
    while (s2 * s2 < c2) s2++;
    if (s2 * s2 == c2) return 1;

    return 0;
}

// Fibonacci range
export function call_range_(n: i64, resultPtr: u32): void {
    if (n < 0) {
        store<u32>(resultPtr, 0);
        store<u32>(resultPtr + 4, 0);
        return;
    }

    const count: u32 = (n + 1) as u32;
    const ptr = alloc(count * 8);

    var a: i64 = 0;
    var b: i64 = 1;
    for (let i: i64 = 0; i <= n; i++) {
        store<i64>(ptr + (i as usize) * 8, a);
        const tmp = a + b;
        a = b;
        b = tmp;
    }

    store<u32>(resultPtr, ptr);
    store<u32>(resultPtr + 4, count);
}
